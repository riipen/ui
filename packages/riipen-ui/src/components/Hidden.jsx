import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class Hidden extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * Given some size, will hide content at that size and up / down (direction).
     */
    direction: PropTypes.oneOf(["down", "up"]),

    /**
     * The size at which to hide content in tandum with a direction.
     */
    size: PropTypes.oneOf(["xs", "sm", "md", "lg", "xl"]).isRequired
  };

  static defaultProps = {
    direction: "down"
  };

  static contextType = ThemeContext;

  render() {
    const { children, direction, size } = this.props;

    const theme = this.context;

    const className = clsx(`${size}-${direction}`);

    return (
      <React.Fragment>
        <div className={className}>{children}</div>
        <style jsx>{`
          @media (min-width: ${theme.breakpoints.xs}px) {
            .xs-up {
              display: none;
            }
          }
          @media (max-width: ${theme.breakpoints.xs}px) {
            .xs-down {
              display: none;
            }
          }

          @media (min-width: ${theme.breakpoints.sm}px) {
            .sm-up {
              display: none;
            }
          }
          @media (max-width: ${theme.breakpoints.sm}px) {
            .sm-down {
              display: none;
            }
          }

          @media (min-width: ${theme.breakpoints.md}px) {
            .md-up {
              display: none;
            }
          }
          @media (max-width: ${theme.breakpoints.md}px) {
            .md-down {
              display: none;
            }
          }

          @media (min-width: ${theme.breakpoints.lg}px) {
            .lg-up {
              display: none;
            }
          }
          @media (max-width: ${theme.breakpoints.lg}px) {
            .lg-down {
              display: none;
            }
          }

          @media (min-width: ${theme.breakpoints.xl}px) {
            .xl-up {
              display: none;
            }
          }
          @media (max-width: ${theme.breakpoints.xl}px) {
            .xl-down {
              display: none;
            }
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Hidden);
