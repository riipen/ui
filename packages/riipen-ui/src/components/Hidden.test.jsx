import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Hidden from "./Hidden.jsx";

describe("<Hidden>", () => {
  it("renders without errors", () => {
    let error;
    const size = "sm";

    try {
      mount(<Hidden size={size} />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const size = "sm";

    const wrapper = mount(<Hidden size={size} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const size = "sm";
      const defaultProps = new Hidden({ size }).type.defaultProps;

      const wrapper = mount(<Hidden size={size} />);

      const component = wrapper.find("Hidden");
      expect(component.props().direction).toEqual(defaultProps.direction);
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <h1>this is a child</h1>;
      const size = "sm";

      const wrapper = mount(<Hidden size={size}>{child}</Hidden>);

      expect(
        wrapper
          .find("Hidden")
          .find("div")
          .contains(child)
      ).toEqual(true);
    });
  });

  describe("direction prop", () => {
    it("applies given direction to the size-direction className on div", () => {
      const direction = "up";
      const size = "sm";
      const className = new RegExp(`.+-${direction}`);

      const wrapper = mount(<Hidden direction={direction} size={size} />);

      expect(
        wrapper
          .find("Hidden")
          .find("div")
          .hasClass(className)
      ).toEqual(true);
    });

    it("gives an error given an invalid direction", () => {
      const direction = "left";
      const size = "sm";

      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Hidden size={size} direction={direction} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("size prop", () => {
    it("applies given size to the size-direction className on div", () => {
      const size = "sm";
      const className = new RegExp(`${size}-.+`);

      const wrapper = mount(<Hidden size={size} />);

      expect(
        wrapper
          .find("Hidden")
          .find("div")
          .hasClass(className)
      ).toEqual(true);
    });

    it("gives an error given an invalid size", () => {
      const size = "xxl";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Hidden size={size} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("classes prop", () => {
    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-hidden"];
      const size = "xs";

      const wrapper = mount(<Hidden size={size} />);

      expect(
        wrapper
          .find("Hidden")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });
});
