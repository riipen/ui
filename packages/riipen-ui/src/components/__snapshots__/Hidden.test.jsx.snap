// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hidden> renders correct snapshot 1`] = `
<withClasses(Hidden)
  size="sm"
>
  <Hidden
    classes={
      Array [
        "riipen",
        "riipen-hidden",
      ]
    }
    direction="down"
    size="sm"
  >
    <div
      className="jsx-679367177 sm-down"
    />
    <JSXStyle
      dynamic={
        Array [
          0,
          0,
          600,
          600,
          960,
          960,
          1380,
          1380,
          1920,
          1920,
        ]
      }
      id="1718998684"
    />
  </Hidden>
</withClasses(Hidden)>
`;
